bitcoin/signature.go

```diff
@@ -173,6 +174,149 @@ func SignatureFromCompact(s string) (Signature, error) {
+	sb := canonicalizeInt(sigS)
+
+	// total length of returned signature is 1 byte for each magic and
+	// length (4 total), plus lengths of r and s
+	length := 4 + len(rb) + len(sb)
+
+	// Header
+	b := make([]byte, 4)
+	b[0] = 0x30
+	b[1] = byte(length)
+
+	// Write R
+	b[2] = 0x02
+	b[3] = byte(len(rb))
+	if _, err := w.Write(b); err != nil {
+		return errors.Wrap(err, "header")
+	}
+	if _, err := w.Write(rb); err != nil {
+		return errors.Wrap(err, "r")
+	}
+
+	// Write S
+	b = make([]byte, 2)
+	b[0] = 0x02
+	b[1] = byte(len(sb))
+	if _, err := w.Write(b); err != nil {
+		return errors.Wrap(err, "header")
+	}
+	if _, err := w.Write(sb); err != nil {
+		return errors.Wrap(err, "s")
+	}
+
+	return nil
+}
+
+func (s *Signature) Deserialize(r io.Reader) error {
+	b := make([]byte, 1)
+	if _, err := r.Read(b); err != nil {
+		return errors.Wrap(err, "header byte")
+	}
+
+	// 0x30
+	if b[0] != 0x30 {
+		return errors.New("Signature missing header byte")
+	}
+
+	// length of remaining message
+	if _, err := r.Read(b); err != nil {
+		return errors.Wrap(err, "length")
+	}
+
+	b = make([]byte, b[0])
+	if _, err := r.Read(b); err != nil {
+		return errors.Wrap(err, "data")
+	}
+
+	// 0x02
+	index := 0
+	if b[index] != 0x02 {
+		return errors.New("Signature missing 1st int marker")
+	}
+	index++
+
+	// Length of R.
+	rLen := int(b[index])
+	// must be positive, must be able to fit in another 0x2, <len> <s>
+	// hence the -3. We assume that the length must be at least one byte.
+	index++
+	if rLen <= 0 || rLen > len(b)-index-3 {
+		return errors.New("Signature has bad R length")
+	}
+
+	// Then R itself.
+	rBytes := b[index : index+rLen]
+	switch err := canonicalPadding(rBytes); err {
+	case errNegativeValue:
+		return errors.New("Signature R is negative")
+	case errExcessivelyPaddedValue:
+		return errors.New("Signature R is excessively padded")
+	}
+
+	s.R.SetBytes(rBytes)
+	index += rLen
+	// 0x02. length already checked in previous if.
+	if b[index] != 0x02 {
+		return errors.New("malformed signature: no 2nd int marker")
+	}
+	index++
+
+	// Length of signature S.
+	sLen := int(b[index])
+	index++
+	// S should be the rest of the string.
+	if sLen <= 0 || sLen > len(b)-index {
+		return errors.New("Signature has bad S length")
+	}
+
+	// Then S itself.
+	sBytes := b[index : index+sLen]
+	switch err := canonicalPadding(sBytes); err {
+	case errNegativeValue:
+		return errors.New("Signature S is negative")
+	case errExcessivelyPaddedValue:
+		return errors.New("Signature S is excessively padded")
+	}
+
+	s.S.SetBytes(sBytes)
+	index += sLen
+
+	// sanity check length parsing
+	if index != len(b) {
+		return fmt.Errorf("Signature has bad final length %v != %v", index, len(b))
+	}
+
+	// Verify also checks this, but we can be more sure that we parsed
+	// correctly if we verify here too.
+	// FWIW the ecdsa spec states that R and S must be | 1, N - 1 |
+	// but crypto/ecdsa only checks for Sign != 0. Mirror that.
+	if s.R.Sign() != 1 {
```

Comment:
This whole block could be reduced to a single "validation" func.

Reply:
Do you mean check the 4 mathematical validity tests in one if statement?

Reply:
Moving that block of checks into a function would drop the overalll func complexity, and allow the "validation/guard" to be testable.

Reply:
oh, I see. the code is duplicated too, so it is a good function.